Core/NPCBarrierGlow.cs

```diff
@@ -0,0 +1,164 @@
+
+namespace StarlightRiver.Core
+{ 
+    public class NPCBarrierGlowSystem : ModSystem
+    {
+        public override void PreUpdateNPCs()
+        {
+            NPCBarrierGlow.anyEnemiesWithBarrier = false;
+        }
+    }
+    public class NPCBarrierGlow : IOrderedLoadable
+    {
+        public static bool anyEnemiesWithBarrier = false;
+
+        public static RenderTarget2D NPCTarget;
+
+        public static RenderTarget2D NPCTargetBehindTiles;
+
+        public static Vector2 oldScreenPos = Vector2.Zero;
+
+        private static Color barrierColor => Color.Cyan;
+
+        public float Priority => 1.1f;
+
+        public void Load()
+        {
+            if (Main.dedServ)
+                return;
+
+            ResizeTarget();
+
+            Main.OnPreDraw += Main_OnPreDraw;
+            On.Terraria.Main.DrawNPCs += DrawBarrierOverlay;
+        }
+
+        public void Unload()
+        {
+            Main.OnPreDraw -= Main_OnPreDraw;
+            On.Terraria.Main.DrawNPCs -= DrawBarrierOverlay;
+        }
+
+        public static void ResizeTarget()
+        {
+            Main.QueueMainThreadAction(() =>
+            {
+                NPCTargetBehindTiles = new RenderTarget2D(Main.graphics.GraphicsDevice, Main.screenWidth, Main.screenHeight);
+                NPCTarget = new RenderTarget2D(Main.graphics.GraphicsDevice, Main.screenWidth, Main.screenHeight);
+            });
+        }
+
+        private void Main_OnPreDraw(GameTime obj)
+        {
+            GraphicsDevice gD = Main.graphics.GraphicsDevice;
+            SpriteBatch spriteBatch = Main.spriteBatch;
+
+            if (Main.gameMenu || Main.dedServ || spriteBatch is null  || NPCTarget is null || NPCTargetBehindTiles is null || gD is null || !anyEnemiesWithBarrier)
+                return;
+
+            oldScreenPos = Main.screenPosition;
+
+            RenderTargetBinding[] bindings = gD.GetRenderTargets();
+            gD.SetRenderTarget(NPCTarget);
+            gD.Clear(Color.Transparent);
+
+            spriteBatch.Begin(default, default, default, default, default, null, Main.GameViewMatrix.ZoomMatrix);
+
+            DrawAllNPCS(spriteBatch, false);
+
+            spriteBatch.End();
+
+            gD.SetRenderTarget(NPCTargetBehindTiles);
+            gD.Clear(Color.Transparent);
+
+            spriteBatch.Begin(default, default, default, default, default, null, Main.GameViewMatrix.ZoomMatrix);
+
+            DrawAllNPCS(spriteBatch, true);
+
+            spriteBatch.End();
+            gD.SetRenderTargets(bindings);
+        }
+
+        private static void DrawAllNPCS(SpriteBatch spriteBatch, bool behindTiles)
+        {
+            for (int i = 0; i < Main.npc.Length; i++)
+            {
+                NPC NPC = Main.npc[i];
+
+                if (NPC.behindTiles != behindTiles)
+                    continue;
+
+                if (NPC.active && NPC.GetGlobalNPC<BarrierNPC>().Barrier > 0)
+                {
+                    if (NPC.ModNPC != null)
```

Comment:
a null check and is check together is redundant, ``is`` checks for null already. Also lets favor using logical operators over chaining conditionals